lib/libc/sys/truncate.2
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)truncate.2	8.1 (Berkeley) 6/4/93
.\" $FreeBSD: releng/11.1/lib/libc/sys/truncate.2 282417 2015-05-04 14:47:00Z jhb $
.\"
.Dd May 4, 2015
.Dt TRUNCATE 2
.Os
.Sh NAME
.Nm truncate ,
.Nm ftruncate
.Nd truncate or extend a file to a specified length
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In unistd.h
.Ft int
.Fn truncate "const char *path" "off_t length"
.Ft int
.Fn ftruncate "int fd" "off_t length"
.Sh DESCRIPTION
The
.Fn truncate
system call
causes the file named by
.Fa path
or referenced by
.Fa fd
to be truncated or extended to
.Fa length
bytes in size.
If the file
was larger than this size, the extra data
is lost.
If the file was smaller than this size,
it will be extended as if by writing bytes
with the value zero.
.Pp
The
.Fn ftruncate
system call causes the file or shared memory object backing the file descriptor
.Fa fd
to be truncated or extended to
.Fa length
bytes in size.
The file descriptor must be a valid file descriptor open for writing.
The file position pointer associated with the file descriptor
.Fa fd
will not be modified.
.Sh RETURN VALUES
.Rv -std
If the file to be modified is not a directory or
a regular file, the
.Fn truncate
call has no effect and returns the value 0.
.Sh ERRORS
The
.Fn truncate
system call
succeeds unless:
.Bl -tag -width Er
.It Bq Er ENOTDIR
A component of the path prefix is not a directory.
.It Bq Er ENAMETOOLONG
A component of a pathname exceeded 255 characters,
or an entire path name exceeded 1023 characters.
.It Bq Er ENOENT
The named file does not exist.
.It Bq Er EACCES
Search permission is denied for a component of the path prefix.
.It Bq Er EACCES
The named file is not writable by the user.
.It Bq Er ELOOP
Too many symbolic links were encountered in translating the pathname.
.It Bq Er EPERM
The named file has its immutable or append-only flag set, see the
.Xr chflags 2
manual page for more information.
.It Bq Er EISDIR
The named file is a directory.
.It Bq Er EROFS
The named file resides on a read-only file system.
.It Bq Er ETXTBSY
The file is a pure procedure (shared text) file that is being executed.
.It Bq Er EFBIG
The
.Fa length
argument was greater than the maximum file size.
.It Bq Er EINVAL
The
.Fa length
argument was less than 0.
.It Bq Er EIO
An I/O error occurred updating the inode.
.It Bq Er EFAULT
The
.Fa path
argument
points outside the process's allocated address space.
.El
.Pp
The
.Fn ftruncate
system call
succeeds unless:
.Bl -tag -width Er
.It Bq Er EBADF
The
.Fa fd
argument
is not a valid descriptor.
.It Bq Er EINVAL
The
.Fa fd
argument
references a file descriptor that is not a regular file or shared memory object.
.It Bq Er EINVAL
The
.Fa fd
descriptor
is not open for writing.
.El
.Sh SEE ALSO
.Xr chflags 2 ,
.Xr open 2 ,
.Xr shm_open 2
.Sh HISTORY
The
.Fn truncate
and
.Fn ftruncate
system calls appeared in
.Bx 4.2 .
.Sh BUGS
These calls should be generalized to allow ranges
of bytes in a file to be discarded.
.Pp
Use of
.Fn truncate
to extend a file is not portable.


ChangeLog
2007-01-25 Christopher Olsen <cwolsen@domainatlantic.com>
	Cleaning up the ne2k driver for future network support. 
	 
	Previous version of UbixOS had LwIP functioning, I hope this will be true in this case. 



lib/libc_bsd/sys/truncate.c
/*
 * Copyright (c) 1992, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)truncate.c	8.1 (Berkeley) 6/17/93";
#endif /* LIBC_SCCS and not lint */
#include <sys/cdefs.h>
__FBSDID("$FreeBSD: releng/10.2/lib/libc/sys/truncate.c 171219 2007-07-04 23:27:38Z peter $");

#include <sys/types.h>
#include <sys/syscall.h>
#include <unistd.h>
#include "libc_private.h"

/*
 * This function provides 64-bit offset padding that
 * is not supplied by GCC 1.X but is supplied by GCC 2.X.
 */
int
truncate(path, length)
	const char	*path;
	off_t	length;
{

	if (__getosreldate() >= 700051)
		return(__sys_truncate(path, length));
	else
		return(__sys_freebsd6_truncate(path, 0, length));
}
